library/src/math/align.rs

```diff
@@ -16,34 +16,30 @@ impl LayoutMath for AlignPointElem {
 
 /// Determine the position of the alignment points.
 pub(super) fn alignments(rows: &[MathRow]) -> Vec<Abs> {
-    let count = rows
-        .iter()
-        .map(|row| {
-            row.iter()
-                .filter(|fragment| matches!(fragment, MathFragment::Align))
-                .count()
-        })
-        .max()
-        .unwrap_or(0);
+    let mut widths = Vec::<Abs>::new();
 
-    let mut points = vec![Abs::zero(); count];
-    for current in 0..count {
-        for row in rows {
-            let mut x = Abs::zero();
-            let mut i = 0;
-            for fragment in row.iter() {
-                if matches!(fragment, MathFragment::Align) {
-                    if i < current {
-                        x = points[i];
-                    } else if i == current {
-                        points[i].set_max(x);
-                    }
-                    i += 1;
+    for row in rows {
+        let mut width = Abs::zero();
+        let mut alignment_index = 0;
+        for fragment in row.iter() {
+            if matches!(fragment, MathFragment::Align) {
+                if alignment_index < widths.len() {
+                    widths[alignment_index].set_max(width);
+                } else {
+                    widths.push(width);
                 }
-                x += fragment.width();
+                width = Abs::zero();
+                alignment_index += 1;
+            } else {
+                width += fragment.width();
             }
         }
     }
 
+    let mut points = widths;
+    for i in 1..points.len() {
+        let prev = points[i - 1];
+        points[i] += prev;
+    }
```

Comment:
Wouldn't this be done better as a cumsum iterator over widths? 

PS. I don't know Rust but this seems like a common pattern that can be done with the local flavour of folds and scans, avoiding mutability.

Reply:
It could, but this way we don't need to pray that the compiler omits the extra allocation it would incur.